Modify PR validation tests to run in a container on openshift

tiwillia opened the pull request:
When a pull request is submitted to the openshift-tools repository on github, a webhook starts a job on a persistent jenkins instance running in an openshift environment. The job is a simple jenkins pipeline using the `Jenkinsfile` in this directory. The pipeline determines the action to take based on the action in the webhook. When a test is necessary, the jenkins pipeline starts a build on the openshift environment titled `openshift-tools-test` and outputs the build logs to the pipeline's logs.

The `openshift-tools-test` build is defined in the `openshift-tools-pr-automation-template.json` template. The build mounts a secret containing the github bot user credentials, then starts a docker build using the Dockerfile `test/Dockerfile`. This build runs from the oso-host-monitoring container and clones the openshift-tools repository. Once the environment is set up, the build runs the `test/run_tests.py` script. This script merges in the changes defined in the github webhook payload and defines several environment variables for consumption by a suite of validators. Each `.py` or .sh` file in `test/validators` is run with these environment varibles.

The script will then submit a comment to the pull request indiciating whether all of the validators passed or if one or more of them failed. The build will exist successfully if all validators pass. When the build completes, so does the jenkins pipeline. The build in jenkins is also marked as a success only if all validators pass.

tiwillia committed:
Modify PR validation tests to run in a container on openshift

tiwillia commented:
If necessary, we can leave the files needed for the current tests instead of deleting them so that tests continue to run while we implement the changes.

tiwillia committed:
Add commit status updates to the test process

A status update will be made when tests start, allowing users to see
that tests are in progress. Then, once tests complete, another status
update will be made on the same commit with the results and a link to
the jenkins build console output.

tiwillia committed:
Remove debugging directive in dockerfile to speed up build

tiwillia committed:
Update jenkins CI readme with new steps

Fix a few minor typos in the Jenkinsfile and testing
scripts. Additionally, change the dockerfile to pull from the correct
openshift-tools repository.

kwoodson commented on jenkins/test/run_tests.py:
```diff
@@ -0,0 +1,200 @@
+            executer = "/usr/bin/python"
+        elif ext == ".sh":
+            executer = "/bin/sh"
+	# If the ext is not recongized, try to just run the file
```
This makes me nervous and we should chat about it.  I won't block for now but I can see this being a black hole.  I believe this would allow someone to run code in our jenkins instance, is that the case?

kwoodson commented on jenkins/test/validators/yaml.sh:
```diff
@@ -0,0 +1,24 @@
+fi
+
+# Download the latest yaml validation script
+curl https://raw.githubusercontent.com/openshift/openshift-ansible/master/git/yaml_validation.py -o jenkins/test/validators/yaml_validation.py 2>&1
```
We recently changed this so that yaml_validation.py is in our repository instead of pulling from openshift-ansible.  We need to make sure that we are doing that new style instead of this as we found a bug.

kwoodson reviewed (commented):
Please review and discuss.

tiwillia committed:
Implement admin whitelist

Due to the nature of dynamically running new tests, it is possible for
malicious code to be run. With this change, only users on the whitelist
will be allowed to run testing on their pull requests.

kwoodson commented on jenkins/test/run_tests.py:
```diff
@@ -173,6 +186,16 @@ def main():
+
+    # Check to ensure the user submitting the changes is on the whitelist
+    user = pull_request["user"]["login"]
+    whitelist = get_user_whitelist)
```
is this correct?  looks like a syntax error.
